shell.go

```diff
@@ -0,0 +1,256 @@
+		cleaned = relPath
+	}
+	if cleaned == "." {
+		return Package{}, fmt.Errorf("invalid package name %q: name resolves to the package store root", name)
+	}
+	if !packageNameRegex.MatchString(cleaned) {
+		return Package{}, fmt.Errorf("invalid package name %q: name doesn't match the regexp %#q", name, packageNameRegex)
+	}
+	pkgFS, err := fs.Sub(p, cleaned)
+	if err != nil {
+		return Package{}, fmt.Errorf("package store %q: unable to open package %q: %v", p.path, name, err)
+	}
+	pkg := Package{
+		FS:   pkgFS,
+		Name: cleaned,
+	}
+	pkg.DirectDependencies, err = p.directDependencies(pkg)
+	if err != nil {
+		return Package{}, fmt.Errorf("package store %q: unable to open package %q: cannot determine package dependencies: %v",
+			p.path, name, err)
+	}
+	return pkg, nil
+}
+
+// directDependencies figures out the immediate dependencies for a given
+// package.
+func (p *PackageStore) directDependencies(pkg Package) ([]Package, error) {
+	installedPkgs, installedHashes, err := p.installedPackages()
+	if err != nil {
+		return nil, err
+	}
+	scanner := newDependencyScanner(installedHashes)
+	err = fs.WalkDir(pkg, ".", func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return fmt.Errorf("walk package store: %v", err)
+		}
+
+		// Don't read directories or symlinks.
+		if !d.Type().IsRegular() {
+			return nil
+		}
+		f, err := pkg.Open(path)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+
+		var size int64
+		if info, err := f.Stat(); err == nil {
+			size = info.Size()
+		}
+		return scanner.scan(f, size)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	deps := make([]Package, 0, 4)
+	for i, found := range scanner.found {
+		if found && installedPkgs[i].Name != pkg.Name {
+			deps = append(deps, installedPkgs[i])
+		}
+	}
+	return deps, err
+}
+
+// installedPackages returns slices of all the installed packages and their
+// hashes. Each package in pkgs maps to its hash at the same index in hashes.
+// Keeping the hashes in a separate slice makes it easier for a
+// dependencyScanner to consume.
+func (p *PackageStore) installedPackages() (pkgs []Package, hashes [][]byte, err error) {
+	entries, err := fs.ReadDir(p, ".")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	pkgs = make([]Package, 0, len(entries))
+	hashes = make([][]byte, 0, len(entries))
+	for _, file := range entries {
+		if !file.IsDir() {
+			continue
+		}
+
+		// Make sure that the directory name looks like a Nix package
+		// and parse its hash.
+		name := file.Name()
+		matches := packageNameRegex.FindStringSubmatch(name)
+		if len(matches) < 2 {
+			continue
+		}
+		hashes = append(hashes, []byte(matches[1]))
+		pkgFS, err := fs.Sub(p, name)
+		if err != nil {
+			return nil, nil, err
+		}
+		pkgs = append(pkgs, Package{
+			FS:   pkgFS,
+			Name: name,
+		})
+	}
+	return pkgs, hashes, nil
+}
+
+// dependencyScanner scans through one or more readers for Nix base32 hashes.
+type dependencyScanner struct {
```

Comment:
I think you should move the dependency scanner into it's own file – leave the main file with mainly the functions that define the PackageStore external API

Reply:
By main file do you mean move `PackageStore` and `Package` into `devbox.go`? I'm also realizing `shell.go` doesn't make sense any more as this evolved. `pkgs.go` might be better.

Reply:
I'd say rename `shell.go` to `pkgs.go` or `store.go` and that's it.

The breaking apart of the dependency scanner into something like `scanner.go` makes more sense if these files were in their own package. As part of `devbox` I think it's fine to leave in one file.